modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,375 @@
+	if size < contractHeaderSize {
+		return 0, nil
+	}
+
+	// Sanity check contract file length.
+	if (size-contractHeaderSize)%crypto.HashSize != 0 {
+		return 0, errors.New("contract file has unexpected length and might be corrupted")
+	}
+	return int((size - contractHeaderSize) / crypto.HashSize), nil
+}
+
+// len returns the number of merkle roots. It should always return the same
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// moveLastCachedSubTreeToUncached deletes the last cached subTree and appends
+// its elements to the uncached roots.
+func (mr *merkleRoots) moveLastCachedSubTreeToUncached() error {
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
+	// Sanity check the number of uncached roots before adding a new one.
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		build.Critical("the number of uncachedRoots is too big. They should've been cached by now")
+	}
+	// Calculate the root offset within the file and write it to disk.
+	rootOffset := fileOffsetFromRootIndex(mr.len())
+	if _, err := mr.file.WriteAt(root[:], rootOffset); err != nil {
+		return err
+	}
+	// Add the root to the unached roots.
+	mr.appendRoot(root)
+
+	// Increment the number of roots.
+	mr.numMerkleRoots++
+	return nil
+}
+
+// root returns the root of the merkle roots.
+func (mr *merkleRoots) root() crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
+	for _, st := range mr.cachedSubTrees {
+		if err := tree.PushSubTree(st.height, st.sum); err != nil {
+			// This should never fail.
+			build.Critical(err)
+		}
+	}
+	for _, root := range mr.uncachedRoots {
+		tree.Push(root)
+	}
+	return tree.Root()
+}
+
+// newRoot returns the root of the merkleTree after appending the newRoot
+// without actually appending it.
+func (mr *merkleRoots) newRoot(newRoot crypto.Hash) crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
+	for _, st := range mr.cachedSubTrees {
+		if err := tree.PushSubTree(st.height, st.sum); err != nil {
+			// This should never fail.
+			build.Critical(err)
+		}
+	}
+	for _, root := range mr.uncachedRoots {
+		tree.Push(root)
+	}
+	// Push the new root.
+	tree.Push(newRoot)
+	return tree.Root()
+}
+
+// merkleRoots reads all the merkle roots from disk and returns them. This is
+// not very fast and should only be used for testing purposes.
+func (mr *merkleRoots) merkleRoots() (roots []crypto.Hash, err error) {
```

Comment:
If this function is only intended for testing, I would move it to an _test.go file, and then probably rename it to `merkleRootsFromDisk()`